pkg/requester/selection/all.go

```diff
@@ -0,0 +1,139 @@
+	ctx = log.Ctx(ctx).With().Str("JobID", job.ID()).Logger().WithContext(ctx)
+
+	nodeIDs, err := s.nodeDiscoverer.FindNodes(ctx, *job)
+	if err != nil {
+		return nil, err
+	}
+	log.Ctx(ctx).Debug().Int("Discovered", len(nodeIDs)).Msg("Found nodes for job")
+
+	rankedNodes, err := s.nodeRanker.RankNodes(ctx, *job, nodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	// filter nodes with rank below 0
+	filteredNodes := lo.Filter(rankedNodes, func(node requester.NodeRank, _ int) bool { return node.Rank >= 0 })
+	log.Ctx(ctx).Debug().Int("Ranked", len(filteredNodes)).Msg("Ranked nodes for job")
+
+	selectedInfos := lo.Map(filteredNodes, func(node requester.NodeRank, _ int) model.NodeInfo { return node.NodeInfo })
+	return selectedInfos, nil
+}
+
+// We asked all nodes that should be capable of running the job to bid, so we
+// want to select all the nodes that said they would run the job.
+func (*allNodeSelector) SelectBids(ctx context.Context, job *model.Job, jobState *model.JobState) (accept, reject []model.ExecutionState) {
+	return jobState.GroupExecutionsByState()[model.ExecutionStateAskForBidAccepted], nil
+}
+
+// We specifically want to run the job on *all* nodes – so we need to only retry
+// executions on the *same* node on which it failed. There would be no point /
+// danger in scheduling the execution to a different node.
+//
+// So we just select all of the nodes that have *only* failed executions (a node
+// that failed and then retried successfully should not be retried again!). This
+// only helps work around transient errors (e.g. something random wrong at the
+// node) and not permanent failures, so there is a limit to the number of times
+// we should retry.
+func (s *allNodeSelector) SelectNodesForRetry(ctx context.Context, job *model.Job, jobState *model.JobState) ([]model.NodeInfo, error) {
+	// failedNodes counts the number of failures that the node has seen so far.
+	// A negative value means the job is successful or in progress.
+	failureCounts := make(map[string]int, len(jobState.Executions))
+	for _, execution := range jobState.Executions {
+		numFails, seen := failureCounts[execution.NodeID]
+		if !seen {
+			numFails = 0
+		}
+
+		if numFails < 0 || !execution.State.IsDiscarded() {
+			// The job is in progress or completed: don't retry this node again.
+			failureCounts[execution.NodeID] = -1
+		} else if execution.State.IsDiscarded() {
+			// The execution failed – increment counter.
+			failureCounts[execution.NodeID] = numFails + 1
+		}
+	}
+
+	failedNodes := make([]string, 0, len(jobState.Executions))
+	for nodeID, failureCount := range failureCounts {
+		if failureCount > 0 && failureCount < maxRetriesPerNode {
+			failedNodes = append(failedNodes, nodeID)
+		}
+	}
+
+	log.Ctx(ctx).Debug().Int("Retryable", len(failedNodes)).Msg("Found nodes to retry")
+	if len(failedNodes) == 0 {
+		return nil, nil
+	}
+
+	foundNodes, err := s.nodeDiscoverer.FindNodes(ctx, *job)
```

Comment:
The `jobState` contains an `ExecutoState` which contains the nodeID for the node running the execution, do we need to re-discover the nodes here? Or could we reuse the value from the execution state?

Reply:
Having this here is for two reasons:

1. Our interface expects a `model.NodeInfo` and this is the only way to get one. However… this interface is only consumed in one place and it only ever uses the node ID. So if we wanted to, we could refactor this to only talk about node IDs. Worth it?
2. This call means that if we need to retry a node but the node has disappeared from the network, we have a chance to throw an error (below). Do we need to do this? Atm without this the job would just sit in `CompletedPartially` if the node never comes back. Is that desirable? This is a more broad question, because atm we'll also end up in that state if the node fails more than the retry limit – it won't be retried and we won't throw an error.

Reply:
Looking at the `FindNode` implementations, the operations look relatively cheap, so I'd be fine just leaving this as it is.